### PR TITLE
Adds UpsamplerGeneric

### LINKED ISSUE #48: Support more chroma subsampling ratios
Implementing a generic integer upsampler will allow us to support most chroma subsampling ratios.

### PR DESCRIPTION
**Pull request by iamrohit7**
Implements a Generic Upsampler with support for integer sampling ratios. Fixes #48

### CODE REVIEW

**Comment by kaksmet** [comment]
When I try to decode the image in servo/servo#13164 using your branch it panics with

`thread '<unnamed>' panicked at 'index out of bounds: the len is 2561 but the index is 2561', src/upsampler.rs:241`.

It would be great to have some tests for this code. Maybe converting some of the images in `tests/reftest/images/mozilla` to use 3x sampling factor using `cjpeg`'s `-sample` option?

**Comment by iamrohit7** [comment]
The tests are still failing, but the visual result looks fine. I haven't included the failing tests.

File: src/upsampler.rs [hunk 1]
```diff
@@ -34,12 +35,14 @@ impl Upsampler {
...
             components: upsampler_components,
+            max_horizontal_sampling_factor: h_max
         })
     }
 
     pub fn upsample_and_interleave_row(&self, component_data: &[Vec<u8>], row: usize, output_width: usize, output: &mut [u8]) {
         let component_count = component_data.len();
-        let mut line_buffer = vec![0u8; output_width + 1];
+        let max_component_width = self.components.iter().map(|c| c.width).max().unwrap();
+        let mut line_buffer = vec![0u8; max_component_width * self.max_horizontal_sampling_factor as usize];
```
Comment (kaksmet) [inline src/upsampler.rs]: Here I would like to cache the result of `components.iter().map(|c| c.size.width as usize * c.horizontal_sampling_factor as usize).max().unwrap()` in Upsampler and use that for line_buffer's size instead.

File: src/upsampler.rs [hunk 2]
```diff
@@ -214,3 +225,25 @@ impl Upsample for UpsamplerH2V2 {
...
+    // Uses nearest neighbor sampling
+    fn upsample_row(&self,
+                    input: &[u8],
+                    input_width: usize,
+                    _input_height: usize,
+                    row_stride: usize,
+                    row: usize,
+                    _output_width: usize,
+                    output: &mut [u8]) {
+        let mut index: usize = 0;
```
Comment (kaksmet) [inline src/upsampler.rs]: Remove the explicit type from index

**Pull request merged**